### DISCUSSION TITLE
Extension uninstall hook

### DISCUSSION THREAD

**Discussion by traeok** [zowe/zowe-explorer-vscode#2861] [Squad Discussion]
Incorporating an [uninstall hook](https://code.visualstudio.com/api/references/extension-manifest#extension-uninstall-hook) in Zowe Explorer and its extenders would provide the following benefits:

- Extenders can use the opportunity to remove any temporary files or settings that they contribute
- We might be able to offer a function in Zowe Explorer API that:
  - evaluates the profile types that an extender contributes and determines which types can be removed
  - performs any other needed cleanup from the Zowe Explorer side to fully remove the extender
- Zowe Explorer can remove any temporary files or settings from VS Code that should not persist after it's uninstalled

Following up on our last stand-up discussion, I'm creating this thread to gather our thoughts on adding this hook and discuss how to make it straightforward for extenders to clean up.

One idea that I had in mind for the uninstall hook is to export a function from Zowe Explorer API, and then have extenders execute it from their uninstall hook:

```json
{
    "scripts": {
        "vscode:uninstall": "node -e \"await import(\"@zowe/zowe-explorer-api\").extenderCleanup()\"" 
    }
}
```

Taking this a step forward, the `extenderCleanup` function could support an optional `callback` parameter to allow extenders to perform extra steps after the API is done with its cleanup steps.

**Comment by JWaters02** [comment]
@traeok Could extenders use this hook as an oppurtunity to remove themselves from the zowe config/schema files? If so, that would be really bad imo, because we (my org) rely on the config files never having entries removed from them when things change (except manually we which avoid doing as much as possible), only things added or changed.

**Comment by AarishMansur** [comment]
@traeok  the feature looks really good imo but while digging in codebases i found some tradeoffs for this which are 

- VS Code's uninstall timing : The uninstall hook runs after VS Code deletes the extensions files so when the hook tries to call `extenderCleanup()` the file it needs to run that is already deleted verified through [Vscode docs ](https://code.visualstudio.com/api/references/extension-manifest#extension-uninstall-hook)

-  Shared Static State : All plugins push their profile types into  one shared static array no name tags  so when FTP is uninstalled Zowe Explorer looks at the list and has no idea which items belong to FTP 

-  No Extender Registry :  When FTP registers its tools (USS, MVS, JES APIs)  they go into a shared storage but again no record of "FTP brought these"  so ZE can't cleanly remove just FTP's contributions without risking unintended side effects
